src/herder/TxSetFrame.cpp

```diff
@@ -136,61 +144,142 @@ TxSetFrame::sortForApply()
+    operator()(TxSetFrame::AccountTransactionQueue const* tx1,
+               TxSetFrame::AccountTransactionQueue const* tx2)
     {
-        if (tx1->getSourceID() == tx2->getSourceID())
-            return tx1->getSeqNum() < tx2->getSeqNum();
-        double fee1 = mAccountFeeMap[tx1->getSourceID()];
-        double fee2 = mAccountFeeMap[tx2->getSourceID()];
-        if (fee1 == fee2)
-            return tx1->getSourceID() < tx2->getSourceID();
-        return fee1 > fee2;
+        if (tx1 == nullptr || tx1->mTxs.empty())
+        {
+            return !tx2->mTxs.empty();
+        }
+        if (tx2 == nullptr || tx2->mTxs.empty())
+        {
+            return false;
+        }
+
+        auto& top1 = tx1->mTxs.front();
+        auto& top2 = tx2->mTxs.front();
+
+        // compare fee/minFee between top1 and top2
+        auto v1 = bigMultiply(top1->getFeeBid(), top2->getMinFee(mHeader));
+        auto v2 = bigMultiply(top2->getFeeBid(), top1->getMinFee(mHeader));
+        if (v1 < v2)
+        {
+            return true;
+        }
+        else if (v1 > v2)
+        {
+            return false;
+        }
+        // use hash of transaction as a tie breaker
+        return lessThanXored(top1->getFullHash(), top2->getFullHash(), mSeed);
     }
 };
 
+std::unordered_map<AccountID, TxSetFrame::AccountTransactionQueue>
+TxSetFrame::buildAccountTxQueues()
+{
+    std::unordered_map<AccountID, AccountTransactionQueue> actTxQueueMap;
+    for (auto& tx : mTransactions)
+    {
+        auto& id = tx->getSourceID();
+        auto it = actTxQueueMap.find(id);
+        if (it == actTxQueueMap.end())
+        {
+            auto d = std::make_pair(id, AccountTransactionQueue{id});
+            auto r = actTxQueueMap.insert(d);
+            it = r.first;
+        }
+        it->second.mTxs.emplace_back(tx);
+    }
+
+    for (auto& am : actTxQueueMap)
+    {
+        // sort each in sequence number order
+        std::sort(am.second.mTxs.begin(), am.second.mTxs.end(), SeqSorter);
+    }
+    return actTxQueueMap;
+}
+
 void
 TxSetFrame::surgePricingFilter(Application& app)
 {
     LedgerTxn ltx(app.getLedgerTxnRoot());
     auto header = ltx.loadHeader();
-    size_t max = header.current().maxTxSetSize;
-    if (mTransactions.size() > max)
-    { // surge pricing in effect!
-        CLOG(WARNING, "Herder")
-            << "surge pricing in effect! " << mTransactions.size();
-
-        // determine the fee ratio for each account
-        map<AccountID, double> accountFeeMap;
-        for (auto& tx : mTransactions)
+
+    bool maxIsOps = header.current().ledgerVersion >= 11;
+
+    size_t opsLeft;
+    {
+        auto maxTxSetSize = header.current().maxTxSetSize;
+        opsLeft = maxIsOps ? maxTxSetSize : (maxTxSetSize * 100);
+    }
+
+    auto curSizeOps = maxIsOps ? sizeOp() : (sizeTx() * 100);
+    size_t surgeOpsCutoff = 0;
+    if (opsLeft >= 100)
+    {
+        surgeOpsCutoff = opsLeft - 100;
```

Comment:
why are we doing that?
If max number of operations is set to 10000 and this txset has 10000 why do we remove at least 100 of them?

Reply:
it doesn't remove anything, limit is `opsLeft`, I removed this code